### ISSUE TITLE
ZIP64 support is broken

### ISSUE THREAD

**Issue by koute** [mvdnes/zip-rs#84]
I have a zip file which `zip` fails to open. Unfortunately I can't share it, however I've tracked the issue down to the zip64 extra field parsing.

Here's how `zip` currently handles it:

```
fn parse_extra_field(file: &mut ZipFileData, data: &[u8]) -> ZipResult<()>
{   
    let mut reader = io::Cursor::new(data);
    
    while (reader.position() as usize) < data.len()
    {
        let kind = reader.read_u16::<LittleEndian>()?;
        let len = reader.read_u16::<LittleEndian>()?;
        match kind
        {
            // Zip64 extended information extra field
            0x0001 => {
                file.uncompressed_size = reader.read_u64::<LittleEndian>()?;
                file.compressed_size = reader.read_u64::<LittleEndian>()?;
                reader.read_u64::<LittleEndian>()?;  // relative header offset
                reader.read_u32::<LittleEndian>()?;  // disk start number
            },
            _ => { reader.seek(io::SeekFrom::Current(len as i64))?; },
        };
    }
    Ok(())
}
```

And here's how `unzip` (the command line tool which properly unpacks my zip file) does it:

```
int getZip64Data(__G__ ef_buf, ef_len)
    __GDEF
    ZCONST uch *ef_buf; /* buffer containing extra field */
    unsigned ef_len;    /* total length of extra field */
{
    unsigned eb_id;
    unsigned eb_len;
    
/*---------------------------------------------------------------------------
    This function scans the extra field for zip64 information, ie 8-byte
    versions of compressed file size, uncompressed file size, relative offset
    and a 4-byte version of disk start number.
    Sets both local header and central header fields.  Not terribly clever,
    but it means that this procedure is only called in one place.
  ---------------------------------------------------------------------------*/

    if (ef_len == 0 || ef_buf == NULL)
        return PK_COOL;

    Trace((stderr,"\ngetZip64Data: scanning extra field of length %u\n",
      ef_len));

    while (ef_len >= EB_HEADSIZE) {
        eb_id = makeword(EB_ID + ef_buf);
        eb_len = makeword(EB_LEN + ef_buf);

        if (eb_len > (ef_len - EB_HEADSIZE)) {
            /* discovered some extra field inconsistency! */
            Trace((stderr,
              "getZip64Data: block length %u > rest ef_size %u\n", eb_len,
              ef_len - EB_HEADSIZE));
            break;
        }
        if (eb_id == EF_PKSZ64) {
        
          int offset = EB_HEADSIZE;

          if (G.crec.ucsize == 0xffffffff || G.lrec.ucsize == 0xffffffff){
            G.lrec.ucsize = G.crec.ucsize = makeint64(offset + ef_buf);
            offset += sizeof(G.crec.ucsize);
          }
          if (G.crec.csize == 0xffffffff || G.lrec.csize == 0xffffffff){
            G.csize = G.lrec.csize = G.crec.csize = makeint64(offset + ef_buf);
            offset += sizeof(G.crec.csize);
          }
          if (G.crec.relative_offset_local_header == 0xffffffff){
            G.crec.relative_offset_local_header = makeint64(offset + ef_buf);
            offset += sizeof(G.crec.relative_offset_local_header);
          }
          if (G.crec.disk_number_start == 0xffff){
            G.crec.disk_number_start = (zuvl_t)makelong(offset + ef_buf);
            offset += sizeof(G.crec.disk_number_start);
          }
        }
          
        /* Skip this extra field block */
        ef_buf += (eb_len + EB_HEADSIZE);
        ef_len -= (eb_len + EB_HEADSIZE);
    }

    return PK_COOL;
} /* end function getZip64Data() */
```

The `unzip`'s logic here is obviously very different.

**Comment by mvdnes** [comment]
I see that it was indeed not entirely correct. I have pushed some changes to master (in commit dbb871245), can you check if that resolves your issues? Otherwise I will need some more details on your zip file.

**Comment by koute** [comment]
Thanks!

I don't have the zip file with me on hand right now, however before going home I did quickly poke and prod it to check why exactly it failed, and in that particular case the issue was that the `header_start` field wasn't updated.

I'm not sure if there is a formal spec for this, but can't we just do exactly what `unzip` does for all of the fields? E.g. something like this: (I actually planned to submit this as a PR, but you were faster. :sweat_smile:)

```
diff --git a/src/read.rs b/src/read.rs
index 3949cf4..6ca5ac5 100644
--- a/src/read.rs
+++ b/src/read.rs
@@ -328,7 +328,7 @@ fn central_header_to_zip_file<R: Read+io::Seek>(reader: &mut R, archive_offset:
     let file_name_length = reader.read_u16::<LittleEndian>()? as usize;
     let extra_field_length = reader.read_u16::<LittleEndian>()? as usize;
     let file_comment_length = reader.read_u16::<LittleEndian>()? as usize;
-    let _disk_number = reader.read_u16::<LittleEndian>()?;
+    let disk_number = reader.read_u16::<LittleEndian>()?;
     let _internal_file_attributes = reader.read_u16::<LittleEndian>()?;
     let external_file_attributes = reader.read_u32::<LittleEndian>()?;
     let mut offset = reader.read_u32::<LittleEndian>()? as u64;
@@ -367,6 +367,7 @@ fn central_header_to_zip_file<R: Read+io::Seek>(reader: &mut R, archive_offset:
         header_start: offset,
         data_start: 0,
         external_attributes: external_file_attributes,
+        disk_number: disk_number as u32
     };
 
     match parse_extra_field(&mut result, &*extra_field) {
@@ -409,10 +410,18 @@ fn parse_extra_field(file: &mut ZipFileData, data: &[u8]) -> ZipResult<()>
         {
             // Zip64 extended information extra field
             0x0001 => {
-                file.uncompressed_size = reader.read_u64::<LittleEndian>()?;
-                file.compressed_size = reader.read_u64::<LittleEndian>()?;
-                reader.read_u64::<LittleEndian>()?;  // relative header offset
-                reader.read_u32::<LittleEndian>()?;  // disk start number
+                if file.uncompressed_size == 0xffffffff {
+                    file.uncompressed_size = reader.read_u64::<LittleEndian>()?;
+                }
+                if file.compressed_size == 0xffffffff {
+                    file.compressed_size = reader.read_u64::<LittleEndian>()?;
+                }
+                if file.header_start == 0xffffffff {
+                    file.header_start = reader.read_u64::<LittleEndian>()?;
+                }
+                if file.disk_number == 0xffff {
+                    file.disk_number = reader.read_u32::<LittleEndian>()?;
+                }
             },
             _ => { reader.seek(io::SeekFrom::Current(len as i64))?; },
         };
@@ -614,6 +623,7 @@ pub fn read_zipfile_from_stream<'a, R: io::Read>(reader: &'a mut R) -> ZipResult
         // We set this to zero, which should be valid as the docs state 'If input came
         // from standard input, this field is set to zero.'
         external_attributes: 0,
+        disk_number: 0
     };
 
     match parse_extra_field(&mut result, &extra_field) {
diff --git a/src/types.rs b/src/types.rs
index a8609b3..c8f31c3 100644
--- a/src/types.rs
+++ b/src/types.rs
@@ -59,6 +59,8 @@ pub struct ZipFileData
     pub data_start: u64,
     /// External file attributes
     pub external_attributes: u32,
+    /// Disk number
+    pub disk_number: u32
 }
 
 impl ZipFileData {
diff --git a/src/write.rs b/src/write.rs
index ffcdfa2..f408254 100644
--- a/src/write.rs
+++ b/src/write.rs
@@ -214,6 +214,7 @@ impl<W: Write+io::Seek> ZipWriter<W>
                 header_start: header_start,
                 data_start: 0,
                 external_attributes: permissions << 16,
+                disk_number: 0
             };
             write_local_file_header(writer, &file)?;
 
```

**Comment by mvdnes** [comment]
A yes, this is largely the same I guess. After some thought, I think I need to take the header offset also in consideration (like you do), otherwise zip files larger than 4GB are not able to be parsed. I will do that soon, as some other things like the `archive_offset` also needs to be taken in consideration.

**Comment by mvdnes** [comment]
I think I have fixed the remaining issues in commits c286a8ba16 and bb38674a14. A test with a large zip64 file containing multiple entries now seems to work correctly.
